File: .skills/openclaw-skills/skills/pete2048/pete2048-travel-smart/src/agents/router.py
# -*- coding: utf-8 -*-
"""
LLM Router Agent - 理解用户意图，分发到对应场景
"""
import json
from typing import Optional
from loguru import logger
from ..config.settings import LLM_CONFIG


class RouterAgent:
    """
    意图识别 + 任务分发

    使用 LLM 理解用户用自然语言描述的场景，
    提取结构化参数，分发到对应场景处理模块。
    """

    SYSTEM_PROMPT = """你是一个智能出行决策助手。用户会描述他遇到的情况，你需要识别属于哪个场景并提取参数。

支持三个场景：
1. highway - 高速堵车，想下高速找吃的/玩的
2. hotel - 途中想找住宿
3. taxi - 景点太堵，想找地方停车打车

用户输入可能是：
- "我在G4高速上，前面10公里堵死了，想找个地方吃饭"
- "我想在附近找个酒店住一晚，明天去故宫"
- "故宫门口太堵了，我在附近找个地方停车打车"

请以JSON格式返回，不要有其他内容：
{
  "scene": "highway|hotel|taxi",
  "params": {
    // highway场景：highway(高速名), location(当前位置描述), destination(目的地描述)
    // hotel场景：location(当前位置描述), budget(预算), people(人数), next_destination(次日目的地)
    // taxi场景：location(堵车位置描述), destination(想去的景点名)
  },
  "reasoning": "一句话说明为什么选择这个场景"
}
"""

    def __init__(self):
        self.llm_config = LLM_CONFIG
        self._client = None

    @property
    def client(self):
        """懒加载 LLM 客户端"""
        if self._client is None:
            import requests
            self._client = requests
        return self._client

    def parse(self, user_input: str) -> Optional[dict]:
        """
        解析用户输入，识别场景和参数

        Returns:
            {"scene": "highway", "params": {...}, "reasoning": "..."}
        """
        try:
            payload = {
                "model": self.llm_config["model"],
                "messages": [
                    {"role": "system", "content": self.SYSTEM_PROMPT},
                    {"role": "user", "content": user_input},
                ],
                "temperature": 0.1,
            }
            headers = {
                "Authorization": f"Bearer {self.llm_config['api_key']}",
                "Content-Type": "application/json",
            }
            resp = self.client.post(
                f"{self.llm_config['base_url']}/v1/chat/completions",
                json=payload,
                headers=headers,
                timeout=30,
            )
            if resp.status_code != 200:
                logger.error(f"LLM 调用失败: {resp.status_code} {resp.text}")
                return None
            result = resp.json()
            content = result["choices"][0]["message"]["content"].strip()
            # 尝试解析 JSON
            if "```json" in content:
                content = content.split("```json")[1].split("```")[0].strip()
            elif "```" in content:
                content = content.split("```")[1].split("```")[0].strip()
            return json.loads(content)
        except Exception as e:
            logger.error(f"解析用户输入失败: {e}")
            return None

    def route(self, user_input: str) -> Optional[str]:
        """
        判断应该路由到哪个场景

        Returns: "highway" | "hotel" | "taxi" | None
        """
        parsed = self.parse(user_input)
        if parsed:
            return parsed.get("scene")
        return None

File: .skills/openclaw-skills/skills/xdd-xund/chinese-llm-router/scripts/router.js
#!/usr/bin/env node
/**
 * Chinese LLM Router - Core Router
 * Routes chat completions to configured Chinese LLM providers
 * All providers use OpenAI-compatible API format
 */

const fs = require('fs');
const path = require('path');
const https = require('https');
const http = require('http');

const CONFIG_FILE = path.join(
  process.env.HOME || process.env.USERPROFILE,
  '.chinese-llm-router', 'config.json'
);

// Model → Provider mapping (auto-built from config)
const MODEL_ALIASES = {
  // DeepSeek
  'deepseek': 'deepseek-chat',
  'deepseek-v3': 'deepseek-chat',
  'deepseek-v3.2': 'deepseek-chat',
  'deepseek-r1': 'deepseek-reasoner',
  // Qwen
  'qwen': 'qwen3-max',
  'qwen3': 'qwen3-max',
  'qwen-thinking': 'qwen3-max-thinking',
  'qwen-coder': 'qwen3-coder-plus',
  'qwen-turbo': 'qwen-turbo-latest',
  // GLM
  'glm': 'glm-5',
  'zhipu': 'glm-5',
  'chatglm': 'glm-5',
  // Kimi
  'kimi': 'kimi-k2.5',
  'moonshot': 'kimi-k2.5',
  'kimi-thinking': 'kimi-k2.5-thinking',
  // Doubao
  'doubao': 'doubao-seed-2.0-pro',
  'doubao-pro': 'doubao-seed-2.0-pro',
  'doubao-lite': 'doubao-seed-2.0-lite',
  'doubao-mini': 'doubao-seed-2.0-mini',
  '豆包': 'doubao-seed-2.0-pro',
  // MiniMax
  'minimax': 'minimax-m2.5',
  // Step
  'step': 'step-3.5-flash',
  '阶跃': 'step-3.5-flash',
  // Baichuan
  'baichuan': 'baichuan4-turbo',
  '百川': 'baichuan4-turbo',
  // Spark
  'spark': 'spark-v4.0-ultra',
  '星火': 'spark-v4.0-ultra',
  // Hunyuan
  'hunyuan': 'hunyuan-turbo-s',
  '混元': 'hunyuan-turbo-s'
};

// Task → Model recommendations
const TASK_RECOMMENDATIONS = {
  'coding': ['glm-5', 'kimi-k2.5', 'qwen3-coder-plus', 'deepseek-chat'],
  'math': ['deepseek-reasoner', 'qwen3-max-thinking', 'kimi-k2.5-thinking'],
  'reasoning': ['deepseek-reasoner', 'qwen3-max-thinking', 'kimi-k2.5-thinking'],
  'writing': ['qwen3-max', 'doubao-seed-2.0-pro', 'kimi-k2.5'],
  'translation': ['qwen3-max', 'baichuan4-turbo', 'deepseek-chat'],
  'chat': ['qwen3-max', 'doubao-seed-2.0-pro', 'deepseek-chat'],
  'fast': ['step-3.5-flash', 'doubao-seed-2.0-mini', 'qwen-turbo-latest'],
  'cheap': ['deepseek-chat', 'doubao-seed-2.0-mini', 'step-3.5-flash'],
  'long': ['kimi-k2.5', 'deepseek-chat', 'qwen3-max'],
  'agent': ['glm-5', 'qwen3-max', 'doubao-seed-2.0-pro'],
  'vision': ['kimi-k2.5', 'qwen3-max', 'glm-5']
};

function loadConfig() {
  if (!fs.existsSync(CONFIG_FILE)) {
    throw new Error(`Config not found. Run: node scripts/setup.js`);
  }
  return JSON.parse(fs.readFileSync(CONFIG_FILE, 'utf8'));
}

function resolveModel(name) {
  const lower = name.toLowerCase().trim();
  return MODEL_ALIASES[lower] || lower;
}

function findProvider(config, modelId) {
  for (const [providerId, provider] of Object.entries(config.providers)) {
    if (provider.models.includes(modelId)) {
      return { providerId, ...provider };
    }
  }
  return null;
}

function chatCompletion(provider, model, messages, options = {}) {
  return new Promise((resolve, reject) => {
    const url = new URL(provider.baseUrl + '/chat/completions');
    const isHttps = url.protocol === 'https:';
    const client = isHttps ? https : http;

    const body = JSON.stringify({
      model,
      messages,
      temperature: options.temperature ?? 0.7,
      max_tokens: options.max_tokens ?? 4096,
      stream: options.stream ?? false,
      ...options.extra
    });

    const req = client.request({
      hostname: url.hostname,
      port: url.port || (isHttps ? 443 : 80),
      path: url.pathname + '/chat/completions',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${provider.apiKey}`,
        'Content-Length': Buffer.byteLength(body)
      }
    }, (res) => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        if (res.statusCode >= 400) {
          reject(new Error(`${res.statusCode}: ${data}`));
        } else {
          try { resolve(JSON.parse(data)); }
          catch (e) { reject(new Error(`Parse error: ${data.slice(0, 200)}`)); }
        }
      });
    });

    req.on('error', reject);
    req.write(body);
    req.end();
  });
}

async function route(modelName, messages, options = {}) {
  const config = loadConfig();
  const modelId = resolveModel(modelName);
  const provider = findProvider(config, modelId);

  if (!provider) {
    // Try fallback
    for (const fb of (config.fallback || [])) {
      const fbProvider = findProvider(config, fb);
      if (fbProvider) {
        console.warn(`⚠️ Model "${modelName}" not configured, falling back to ${fb}`);
        return chatCompletion(fbProvider, fb, messages, options);
      }
    }
    throw new Error(`Model "${modelName}" not found. Run: node scripts/setup.js`);
  }

  return chatCompletion(provider, modelId, messages, options);
}

function listModels() {
  const config = loadConfig();
  const result = [];
  for (const [id, provider] of Object.entries(config.providers)) {
    for (const model of provider.models) {
      result.push({ provider: id, model, baseUrl: provider.baseUrl });
    }
  }
  return result;
}

function recommend(task) {
  const config = loadConfig();
  const recs = TASK_RECOMMENDATIONS[task.toLowerCase()] || TASK_RECOMMENDATIONS['chat'];
  return recs.filter(m => findProvider(config, m)).map(m => ({
    model: m,
    provider: findProvider(config, m)?.providerId
  }));
}

// CLI mode
if (require.main === module) {
  const args = process.argv.slice(2);
  const cmd = args[0];

  if (cmd === 'list') {
    const models = listModels();
    console.log('\n🇨🇳 Available Chinese LLMs:\n');
    for (const m of models) {
      console.log(`  ${m.model} (${m.provider})`);
    }
    console.log(`\nTotal: ${models.length} models from ${new Set(models.map(m => m.provider)).size} providers`);
  } else if (cmd === 'recommend') {
    const task = args[1] || 'chat';
    const recs = recommend(task);
    console.log(`\n🎯 Recommended for "${task}":\n`);
    recs.forEach((r, i) => console.log(`  ${i + 1}. ${r.model} (${r.provider})`));
  } else if (cmd === 'test') {
    const model = args[1] || 'qwen3-max';
    console.log(`\n🧪 Testing ${model}...`);
    route(model, [{ role: 'user', content: '你好，请用一句话介绍你自己。' }])
      .then(r => {
        console.log(`✅ ${model}: ${r.choices[0].message.content}`);
        if (r.usage) console.log(`   Tokens: ${r.usage.prompt_tokens} in / ${r.usage.completion_tokens} out`);
      })
      .catch(e => console.error(`❌ ${model}: ${e.message}`));
  } else {
    console.log('Usage: node router.js <command>');
    console.log('  list              - List all configured models');
    console.log('  recommend <task>  - Get model recommendation (coding/math/writing/fast/cheap/...)');
    console.log('  test <model>      - Test a model with a simple prompt');
  }
}

module.exports = { route, listModels, recommend, resolveModel, MODEL_ALIASES, TASK_RECOMMENDATIONS };
